### PR TITLE
CKAPPS-3912 by rlm: Allow checklist item handlers to be context produ…

### PR DESCRIPTION
**Pull request by rlmumford**
…cers and consumers.

### CODE REVIEW

File: modules/data/task/contrib/job/src/Controller/ChooseHandlerController.php [hunk 1]
```diff
@@ -42,13 +79,27 @@ public static function create(ContainerInterface $container) {
    *   The checklist item handler manager.
    * @param \Drupal\Core\Form\FormBuilderInterface $form_builder
    *   The form builder service.
+   * @param \Drupal\Core\Plugin\Context\ContextHandlerInterface $context_handler
+   *   The context handler service.
+   * @param \Drupal\checklist\ChecklistContextCollectorInterface $context_collector
+   *   The checklist context collector service.
+   * @param \Drupal\checklist\ChecklistTypeManager $checklist_type_manager
+   *   The checklist type manager.
    */
```
Comment (jessarexward) [inline modules/data/task/contrib/job/src/Controller/ChooseHandlerController.php:88]: add tempstore repo param comment

File: modules/data/task/contrib/job/src/JobConfigurationChecklist.php [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+namespace Drupal\task_job;
+
+use Drupal\checklist\Checklist;
+use Drupal\checklist\Plugin\ChecklistType\ChecklistTypeInterface;
+use Drupal\Component\Plugin\PluginManagerInterface;
+use Drupal\task\Entity\Task;
+use Drupal\task_job\Plugin\ChecklistType\Job;
+
+/**
+ * A checklist object for when configuring a job.
```
Comment (jessarexward) [inline modules/data/task/contrib/job/src/JobConfigurationChecklist.php:12]: for configuring *

File: modules/data/task/contrib/job/src/JobConfigurationChecklist.php [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+  ) {
+    if (!($type instanceof Job)) {
+      throw new \InvalidArgumentException('Only job configuration types are compatible with ' . static::class);
+    }
+
+    /** @var \Drupal\task_job\Plugin\ChecklistType\Job $type */
+    parent::__construct(
+      $type,
+      Task::create([
+        'job' => $type->getJob(),
```
Comment (jessarexward) [inline modules/data/task/contrib/job/src/JobConfigurationChecklist.php:56]: are there any other fields that are required to create a task entity? I think the title and description are

File: modules/data/checklist/src/Checklist.php [hunk 4]
```diff
@@ -168,6 +170,22 @@ public function removeItem(string $name) {
     unset($this->items[$name]);
   }
 
+  /**
+   * {@inheritdoc}
+   */
+  public function setItem(string $name, ChecklistItemInterface $item) {
+    if ($item->name->isEmpty()) {
+      $item->name = $name;
```
[the remark this answers is not in the document]
Comment (rlmumford) [reply modules/data/checklist/src/Checklist.php:178]: I don't think so, the item only needs to be saved when the status on it changes or something. It's better to leave that up to the caller of setItem.

File: modules/data/checklist/src/Event/ChecklistEvent.php [hunk 5]
```diff
@@ -3,7 +3,7 @@
 namespace Drupal\checklist\Event;
 
 use Drupal\checklist\ChecklistInterface;
-use Symfony\Component\EventDispatcher\Event;
+use Drupal\Component\EventDispatcher\Event;
```
[the remark this answers is not in the document]
Comment (rlmumford) [reply modules/data/checklist/src/Event/ChecklistEvent.php:6]: The symfony class is deprecated, Drupal is providing a BC layer.

File: modules/data/checklist/src/Event/ChecklistItemEvent.php [hunk 6]
```diff
@@ -3,7 +3,7 @@
 namespace Drupal\checklist\Event;
 
 use Drupal\checklist\Entity\ChecklistItemInterface;
-use Symfony\Component\EventDispatcher\Event;
+use Drupal\Component\EventDispatcher\Event;
```
[the remark this answers is not in the document]
Comment (rlmumford) [reply modules/data/checklist/src/Event/ChecklistItemEvent.php:6]: The symfony class is deprecated, Drupal is providing a BC layer.

File: modules/data/checklist/src/EventSubscriber/ChecklistContextsEventSubscriber.php [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+ * Subscriber for checklist contexts.
+ */
+class ChecklistContextsEventSubscriber implements EventSubscriberInterface {
+  use TypedDataDefinitionToContextDefinitionTrait;
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function getSubscribedEvents() {
+    $events[ChecklistEvents::COLLECT_RUNTIME_CONTEXTS][] = ['addChecklistEntityContext', 255];
```
[the remark this answers is not in the document]
Comment (rlmumford) [reply modules/data/checklist/src/EventSubscriber/ChecklistContextsEventSubscriber.php]: The priority of the listener. 255 is the max priority, so this runs first.

File: modules/data/checklist/src/EventSubscriber/ChecklistContextsEventSubscriber.php [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+        $event->addContext(
+          "item:{$name}:{$outcome_name}",
+          new Context($this->contextDefinitionForDataDefinition($definition))
+        );
+      }
+    }
+  }
+
+  /**
+   * Get the runtime contexts from actual outcomes.
```
[the remark this answers is not in the document]
Comment (rlmumford) [reply modules/data/checklist/src/EventSubscriber/ChecklistContextsEventSubscriber.php:86]: Done

File: modules/data/checklist/src/Plugin/ChecklistItemHandler/CreateEntity.php [hunk 9]
```diff
@@ -194,4 +197,25 @@ public function getEntityType() : EntityTypeInterface {
...
 
+  /**
+   * {@inheritdoc}
+   */
+  public function expectedOutcomeDefinitions(): array {
+    $conf = $this->getConfiguration();
+
+    $definition = EntityDataDefinition::create(
+      $this->entityType->id(),
+      $conf['bundle'] !== '__select' ? $conf['bundle'] : NULL
```
[the remark this answers is not in the document]
Comment (rlmumford) [reply modules/data/checklist/src/Plugin/ChecklistItemHandler/CreateEntity.php:208]: I'm assuming you meant around the comparison?  If so, done

File: modules/data/checklist/src/Plugin/ChecklistItemHandler/UpdateEntity.php [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+  public function buildConfigurationSummary(): array {
+    $build = parent::buildConfigurationSummary();
+
+    $conf = $this->getConfiguration();
+    $build['form_mode'] = [
+      '#type' => 'item',
+      '#title' => new TranslatableMarkup('Form Mode'),
+      '#markup' => $conf['form_mode'] ?
+        $this->entityDisplayRepository->getFormModeOptions($this->entityType->id())[$conf['form_mode']] :
+        'Unknown'
```
[the remark this answers is not in the document]
Comment (rlmumford) [reply modules/data/checklist/src/Plugin/ChecklistItemHandler/UpdateEntity.php]: I went with "Not configured"?

File: modules/data/checklist/src/Form/ChecklistItemActionForm.php [hunk 11]
```diff
@@ -47,6 +55,20 @@ class ChecklistItemActionForm extends ChecklistItemFormBase {
    */
   protected $renderer;
 
+  /**
+   * The context handler service.
+   *
+   * @var \Drupal\Core\Plugin\Context\ContextHandlerInterface
+   */
+  protected ContextHandlerInterface $contextHandler;
```
[the remark this answers is not in the document]
Comment (rlmumford) [reply modules/data/checklist/src/Form/ChecklistItemActionForm.php:57]: done

File: modules/data/task/contrib/job/src/Controller/ChooseHandlerController.php (hunk 1 above)
Comment (rlmumford) [reply modules/data/task/contrib/job/src/Controller/ChooseHandlerController.php:90]: done

File: modules/data/task/contrib/job/src/JobConfigurationChecklist.php (hunk 2 above)
Comment (rlmumford) [reply modules/data/task/contrib/job/src/JobConfigurationChecklist.php:12]: done

File: modules/data/task/contrib/job/src/JobConfigurationChecklist.php (hunk 3 above)
Comment (rlmumford) [reply modules/data/task/contrib/job/src/JobConfigurationChecklist.php:56]: Those are required to save, but not to create.

**Pull request merged**